lib/singleton.rb

```diff
@@ -112,40 +112,31 @@ def _dump(depth = -1)
   module SingletonClassMethods # :nodoc:
 
     def clone # :nodoc:
-      Singleton.__init__(super)
+      super.include(Singleton)
     end
 
     # By default calls instance(). Override to retain singleton state.
-    def _load(str)
+    def _load(*)
       instance
     end
 
     def instance # :nodoc:
-      return @singleton__instance__ if @singleton__instance__
-      @singleton__mutex__.synchronize {
-        return @singleton__instance__ if @singleton__instance__
-        @singleton__instance__ = new()
-      }
+      unless @singleton__instance__
+        @singleton__mutex__.synchronize { @singleton__instance__ ||= new }
+      end
+
       @singleton__instance__
     end
 
     private
 
     def inherited(sub_klass)
       super
-      Singleton.__init__(sub_klass)
+      sub_klass.include(Singleton)
```

Comment:
This isn't safe if `super` here copies instance variables from the superclass to the subclass, for similar reasons as for `clone`.